Boundary surface orientation synchronisation

asalzburger opened the pull request:
This is the first part of #185, it synchronises the boundary surface orientation (for the moment left the GenericCuboid untouched), and removes the `BoundaryOrientation` in favour of `NavigationDirection`, which were identical enums.

It renames inside/outside to along/opposite, which is the orientation w.r.t normal vector.

It closes #73.

asalzburger commented on Core/include/Acts/Geometry/CuboidVolumeBounds.hpp:
```diff
@@ -99,11 +98,18 @@ class CuboidVolumeBounds : public VolumeBounds {
+  ///
+  /// @return a vector of surfaces bounding this volume
+  OrientedSurfaces orientedSurfaces(
+      const Transform3D* transform = nullptr) const override;
```
`orientedSurface(...)` replaces `decomposeToSurfaces(...)`

asalzburger commented on Core/include/Acts/Geometry/CutoutCylinderVolumeBounds.hpp:
```diff
@@ -35,6 +36,7 @@ class IVisualization;
 /// --------- hlZ -------
 ///
 ///
+/// @todo add sectoral cutouts
```
Currently, these do not support sectoral descriptions, will add this.

asalzburger commented on Core/include/Acts/Geometry/CylinderVolumeHelper.hpp:
```diff
@@ -6,10 +6,6 @@
-
 #pragma once
 
 #ifndef ACTS_TOOLS_TAKESMALLERBIGGER
```
These macros will go.

asalzburger replied:
That's followed up in #199.

asalzburger commented on Core/src/Geometry/AbstractVolume.cpp:
```diff
@@ -32,25 +28,22 @@ Acts::AbstractVolume::boundarySurfaces() const {
-  for (auto& sf : surfaces) {
-    // flip inner/outer for cylinders
-    AbstractVolume* inner =
-        (sf->type() == Surface::Cylinder && sfCounter == 3 && sfNumber > 3)
```
The special treatment for the 4th surface if Cylinder is gone.

asalzburger commented:
I might have to push again here.

FabianKlimpel commented on Tests/UnitTests/Core/Geometry/GenericCuboidVolumeBoundsTests.cpp:
```diff
@@ -52,7 +52,7 @@ BOOST_AUTO_TEST_CASE(construction_test) {
 }
 
-BOOST_AUTO_TEST_CASE(decomposeToSurfaces) {
+BOOST_AUTO_TEST_CASE(BenericCuboidBoundsOrientedSurfaces) {
```
```suggestion
BOOST_AUTO_TEST_CASE(GenericCuboidBoundsOrientedSurfaces) {
```

asalzburger replied:
Fixed.

asalzburger commented on Core/include/Acts/Geometry/BoundarySurfaceT.hpp:
```diff
@@ -24,48 +20,48 @@ class Surface;
 ///
-/// @note inside/outside definitions are given by the normal vector of the
-/// surface
+/// @note along/oppose definitions are given with respet to the normal vector
```
Fixed.

asalzburger commented on Core/include/Acts/Geometry/BoundarySurfaceT.hpp:
```diff
@@ -24,48 +20,48 @@ class Surface;
+/// @todo change to one schema with BinnedArray0D
 
-template <class T>
+template <class volume_t>
```
Yes, a boundary surface of a `TrackingVolume` is different from one of an `AbstractVolume`.

asalzburger commented on Core/include/Acts/Geometry/BoundarySurfaceT.hpp:
```diff
@@ -24,48 +20,48 @@ class Surface;
-  BoundarySurfaceT(std::shared_ptr<const Surface> surface, const T* inside,
-                   const T* outside)
+  BoundarySurfaceT(std::shared_ptr<const Surface> surface,
+                   const volume_t* inside, const volume_t* outside)
```
They have to be of the same type, as the Navigator goes from `volume_t` to `volume_t`.

FabianKlimpel replied:
Why do they have to be the same? This is just our implementation but in general it shouldn't matter what shape my volumes have

asalzburger commented on Core/include/Acts/Geometry/CylinderVolumeBounds.hpp:
```diff
@@ -238,8 +244,8 @@ inline double CylinderVolumeBounds::binningBorder(BinningValue bValue) const {
-template <class T>
-T& CylinderVolumeBounds::dumpT(T& tstream) const {
+template <class stream_t>
+stream_t& CylinderVolumeBounds::dumpT(stream_t& tstream) const {
```
Will have a look.

asalzburger replied:
Fixed.

asalzburger commented on Core/include/Acts/Geometry/CuboidVolumeBounds.hpp:
```diff
@@ -125,15 +131,20 @@ class CuboidVolumeBounds : public VolumeBounds {
+  /// @tparam stream_t The type fo the dump stream
+  /// @param dt The dump stream object
+  template <class stream_t>
+  stream_t& dumpT(stream_t& dt) const;
```
That's historic, we can streamline this in another MR.

asalzburger commented on Core/src/Geometry/CylinderVolumeBounds.cpp:
```diff
@@ -54,60 +54,61 @@ Acts::CylinderVolumeBounds::CylinderVolumeBounds(
+      Surface::makeShared<CylinderSurface>(trfShared, m_outerCylinderBounds);
+  oSurfaces.push_back(OrientedSurface(std::move(cSurface), backward));
+
+  // [3] Inner Cylinder (optional)
```
Because in the `CutoutCylinderVolumeBounds` the access is with `.at()` while here it's *just* a `.push_back(..)`.

FabianKlimpel replied:
I was more concerned about unnecessary memory allocation

asalzburger commented on Tests/UnitTests/Core/Geometry/CutoutCylinderVolumeBoundsTests.cpp:
```diff
@@ -192,12 +193,34 @@ BOOST_AUTO_TEST_CASE(CutoutCylinderVolumeBoundsBoundingBox) {
   ObjTestWriter::writeObj(tPolyhedrons);
 }
 
+BOOST_AUTO_TEST_CASE(CutoutCylinderVolumeOrientedBoundaries) {
```
That would only "NOT" check two surfaces, as the other ones remain unchanged.

FabianKlimpel replied:
Sure but the corresponding branches are otherwise not covered

asalzburger commented:
I think I pretty much addressed all comments:

- [x]  years updated
- [x] new set of tests for orientation
- [x] new unit test for `VolumeBounds` checking if base orientations are correct

asalzburger commented on Core/include/Acts/Geometry/VolumeBounds.hpp:
```diff
@@ -24,22 +24,30 @@ class VolumeBounds;
+using OrientedSurfaces = std::vector<OrientedSurface>;
+
+// Planar definitions to help construct the boundary surfaces
+static const Transform3D s_planeXY = Transform3D::Identity();
```
I had a `GeometryStatics` in the past, I think we can re-create on when needed?

FabianKlimpel replied:
We will, at a certain point, have this implemented at several places (just a forecast;) )
